### PR TITLE
nvdimm: add case for pmem and alignsize and readonnly

### PR DESCRIPTION
**Pull request by dzhengfy**
Test scenario: support pmem, alignsize and readonly in nvdimm memory device
RHEL-160330

Depend on https://github.com/avocado-framework/avocado-vt/pull/3401

Signed-off-by: Dan Zheng <dzheng@redhat.com>

### CODE REVIEW

**Comment by dzhengfy** [comment]
(1/1) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.check_life_cycle: PASS (119.66 s)
 (1/1) type_specific.io-github-autotest-libvirt.nvdimm.back_file.no_label: PASS (563.71 s)
 (1/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.default: PASS (102.23 s)
 (3/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.discard: PASS (9.21 s)
 (4/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.pmem_alignsize: PASS (55.86 s)
 (6/6) type_specific.io-github-autotest-libvirt.nvdimm.hot_plug: FAIL: error: Failed to detach device from /var/tmp/xml_utils_temp_rtg9kbsa.xml\nerror: internal error: unable to execute QEMU command 'device_del': nvdimm device hot unplug is not supported yet.\n (62.96 s) --- as expected

libvirt-8.2.0-1.el9.x86_64

File: libvirt/tests/src/memory/nvdimm.py [hunk 1]
```diff
@@ -19,107 +20,446 @@
...
+    cmd = 'mount -o dax /dev/pmem0 /mnt'
+    output = vm_session.cmd_output(cmd)
+    test.log.debug("Mount output:%s", output)
+
+    cmd = 'echo \"%s\" >/mnt/%s' % (test_str, test_file)
+    vm_session.cmd(cmd)
+
+    cmd = 'cat /mnt/%s' % test_file
+    if test_str not in vm_session.cmd(cmd):
+        test.fail('"%s" should be in the output of cat cmd' % test_str)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/nvdimm.py:234]: Should we reuse check_nvdimm_file()?

File: libvirt/tests/src/memory/nvdimm.py [hunk 2]
```diff
@@ -19,107 +20,446 @@
...
+
+    # Shutdown the guest, then start it, remount /dev/pmem0,
+    # check if the test file is still on the file system
+    vm.destroy()
+    vm.start()
+    vm_session = vm.wait_for_login()
+
+    vm_session.cmd('mount -o dax /dev/pmem0 /mnt')
+    if test_str not in vm_session.cmd('cat /mnt/foo'):
+        test.fail('\"%s\" should be in /mnt/foo' % test_str)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/nvdimm.py:262]: Same as above?

File: libvirt/tests/src/memory/nvdimm.py [hunk 3]
```diff
@@ -19,107 +20,446 @@
...
+    if IS_PPC_TEST:
+        libvirt.check_qemu_cmd_line('mem-path=/tmp/nvdimm,share=no')
+
+    private_str = 'This is a test for foo-private'
+    vm_session.cmd('mount -o dax /dev/pmem0 /mnt/')
+
+    file_private = 'foo-private'
+    vm_session.cmd("echo '%s' >/mnt/%s" % (private_str, file_private))
+    if private_str not in vm_session.cmd('cat /mnt/%s' % file_private):
+        test.fail('"%s" should be in output' % private_str)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/nvdimm.py:296]: Same as above.

**Review by Yingshun — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/memory/nvdimm.py (hunk 2 above)
Comment (dzhengfy) [reply libvirt/tests/src/memory/nvdimm.py:262]: Updated

File: libvirt/tests/src/memory/nvdimm.py (hunk 1 above)
Comment (dzhengfy) [reply libvirt/tests/src/memory/nvdimm.py:234]: Yes, I missed it. Thanks. Updated.

File: libvirt/tests/src/memory/nvdimm.py (hunk 3 above)
Comment (dzhengfy) [reply libvirt/tests/src/memory/nvdimm.py:296]: updated

Comment (dzhengfy) [reply libvirt/tests/src/memory/nvdimm.py]: Yes, thanks for your carefulness. Updated.

**Review by Yingshun — APPROVED** [review]

**Comment by dzhengfy** [comment]
(1/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.default: PASS (106.55 s)
 (2/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.check_life_cycle: ERROR: Command '/usr/bin/virsh snapshot-create-as avocado-vt-vm1 avocado-vt-vm1.s1' failed.\nstdout: b'\n'\nstderr: b'error: Operation not supported: internal snapshots of a VM with pflash based firmware are not supported\n'\nadditional_info: None (114.47 s) - need to skip the case because internal snapshot is not supported any more in future
 (3/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.discard: PASS (6.74 s)
 (4/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.label.pmem_alignsize: PASS (51.52 s)
 (5/6) type_specific.io-github-autotest-libvirt.nvdimm.back_file.no_label: PASS (552.54 s)
 (6/6) type_specific.io-github-autotest-libvirt.nvdimm.hot_plug: FAIL: error: Failed to detach device from /var/tmp/xml_utils_temp_vt9rgtmk.xml\nerror: internal error: unable to execute QEMU command 'device_del': nvdimm device hot unplug is not supported yet.\n (59.84 s) - as expected

libvirt-8.3.0-1.el9.x86_64

**Review by cliping — APPROVED** [review]
LGTM

**Pull request merged**